### PR TITLE
Add rawType attribute to define binary data for casting the data type from integer to float

### LINKED ISSUE #78: Add rawType attribute to define device reading binary data type
Some device will store data as int16 instead of float32, for example:
1. device read the **temperature 26.3** °C
2. device multiply 26.3 by 100, then convert the result **2630** to binary and store into the register.
image

On the device service side, we should treat the binary as int16 data type and we need to divide 2630 by 100, then the final result should be 26.3 and return to core-data, **but the actual result is 26** because we define the int16 as the properties value type.

Solution:
1. Add rawType attribute( INT16, INT32, ...etc)
2. If rawType exist, device-modbus parse the binary data as rawType and then cast the value according to properties value type.

In this case, rawType is INT16, value type is float32, then device-modbus will parse the binary reading data as INT16 and then cast to float32.

### PR DESCRIPTION
**Pull request by weichou1229**
Fix #78 

Add rawType attribute to define binary data type
* Read command can use rawType to parse the binary data and use valueType to cast the integer to float. 
* Write command can use rawType to cast the float to integer.

This feature only support :
* integer 16 bits to float.
* float to integer 16 bits

### CODE REVIEW

File: internal/driver/deviceclient.go [hunk 1]
```diff
@@ -64,7 +71,26 @@ func createCommandInfo(req *sdkModel.CommandRequest) *CommandInfo {
 		Length:          length,
 		IsByteSwap:      isByteSwap,
 		IsWordSwap:      isWordSwap,
+		rawType:         rawType,
+	}, nil
+}
+
+func getRawType(req *sdkModel.CommandRequest) (valueType sdkModel.ValueType, err error) {
+	rawType, ok := req.Attributes[RAW_TYPE]
+	if !ok {
```
Comment (cloudxxx8) [inline internal/driver/deviceclient.go]: if !ok || rawType == "" {

File: internal/driver/deviceclient.go [hunk 2]
```diff
@@ -79,64 +105,69 @@ func calculateAddressLength(primaryTable string, valueType sdkModel.ValueType) u
...
+			result, err = sdkModel.NewFloat32Value(req.DeviceResourceName, resTime, float32(res))
+			driver.Logger.Debug(fmt.Sprintf("According to the rawType %s and the value type %s, convert integer %d to float %v ", UINT16, FLOAT32, res, result.ValueToString(contract.ENotation)))
+		}
 	case sdkModel.Float64:
-		var res = binary.BigEndian.Uint64(dataBytes)
-		var floatResult = math.Float64frombits(res)
-		result, err = sdkModel.NewFloat64Value(req.DeviceResourceName, resTime, floatResult)
-		stringResult = fmt.Sprint(floatResult)
-
+		if commandInfo.rawType == sdkModel.Float64 {
```
Comment (cloudxxx8) [inline internal/driver/deviceclient.go]: use Switch instead of if else

File: internal/driver/deviceclient.go [hunk 3]
```diff
@@ -79,64 +105,69 @@ func calculateAddressLength(primaryTable string, valueType sdkModel.ValueType) u
...
 		var res = int64(binary.BigEndian.Uint64(dataBytes))
 		result, err = sdkModel.NewInt64Value(req.DeviceResourceName, resTime, res)
-		stringResult = fmt.Sprint(res)
-
 	case sdkModel.Float32:
-		var res = binary.BigEndian.Uint32(dataBytes)
-		var floatResult = math.Float32frombits(res)
-		result, err = sdkModel.NewFloat32Value(req.DeviceResourceName, resTime, floatResult)
-		stringResult = fmt.Sprint(floatResult)
+		if commandInfo.rawType == sdkModel.Float32 {
```
Comment (cloudxxx8) [inline internal/driver/deviceclient.go]: use Switch instead of if else

Comment (weichou1229) [reply internal/driver/deviceclient.go]: Modified.

**Pull request merged**